# Excepciones
---

## Que son las excepciones

- Las características de control de excepciones del lenguaje C# nos ayudan a afrontar cualquier situación inesperada o excepcional que se produce cuando se ejecuta un programa. El control de excepciones usa las palabras clave ***try***, ***catch*** y ***finally*** para intentar realizar acciones que pueden no completarse correctamente, para controlar errores cuando se decide que es razonable hacerlo y para limpiar recursos más adelante.

Las excepciones se crean mediante el uso de la palabra clave ***throw***.

En muchos casos, una excepción la puede no producir un método al que el código ha llamado directamente, sino otro método más bajo en la pila de llamadas. En resumen, el error puede no ser nuestro codigo, por ejemplo que no encuentre un archivo, eso no seria nuestro codigo y podria venir de otro lado el fallo. 
Cuando se genera una excepción, CLR desenreda la pila, busca un método con un bloque catch para el tipo de excepción específico y ejecuta el primer bloque catch que encuentra. Si no encuentra ningún bloque catch adecuado en cualquier parte de la pila de llamadas, finalizará el proceso y mostrará un mensaje al usuario.

En este ejemplo, un método prueba a hacer la división entre cero y detecta el error. Sin el control de excepciones, este programa finalizaría con un error DivideByZeroException no controlada.

```cs

static double SafeDivision(double x, double y) // 2 - Realiza la funcion
{
    if (y == 0) // 3 - Si se da este caso (En el ejemplo SI)
        throw new DivideByZeroException(); // 4 - Ejecuta la parte del Catch
    return x / y;
}

public static void Main()
{
    // Las entradas son para el test. Podemos cambiar los valores, para ver que pasa.
    // comportamiento de las excepciones.
    double a = 98, b = 0;
    double result;

    try
    {
        result = SafeDivision(a, b); // 1 - Lanzamos la funcion
        Console.WriteLine($"{a} divided by {b} = {result}");
    }
    catch (DivideByZeroException) // 5 - Suelta el codigo de aqui dentro. 
    {
        Console.WriteLine("Intentado dividir por zero.");
    }
}
```

El control de excepciones se utiliza sobretodo en casos en el que el programa podria fallar, y no queremos que se detenga su ejecucion, si no que queremos que suelte el error, y continue si puede su ejecucion. Sin el control de excepciones esto nos seria imposible.